File: src/main/java/templemore/onx/version1/Token.java
package templemore.onx.version1;

/**
 * @author Chris Turner
 */
public enum Token {

    nought("O"),
    cross("X");

    private final String symbol;

    Token(String symbol) {
        this.symbol = symbol;
    }

    public String getSymbol() {
        return symbol;
    }

    public Token flip() {
        if ( this == nought ) return cross;
        else return nought;
    }

    @Override
    public String toString() {
        return symbol;
    }
}

File: src/main/scala/templemore/onx/version4/Token.scala
package templemore.onx.version4

/**
 * @author Chris Turner
 */
trait Token {
  def flip(): Token
  def symbol(): String
}

case object Nought extends Token {
  def flip() = Cross
  def symbol() = "O"
}

case object Cross extends Token {
  def flip() = Nought
  def symbol() = "X"
}

